cardano-api/src/Cardano/Api/Tx.hs

```diff
@@ -307,11 +308,16 @@ makeSignedTransaction witnesses (ShelleyTxBody txbody txmetadata) =
                         | ShelleyScriptWitness sw <- witnesses ])
           (Set.fromList [ w | ShelleyBootstrapWitness w <- witnesses ]))
         (maybeToStrictMaybe txmetadata)
+makeSignedTransaction _ (ShelleyTxBody ShelleyBasedEraAllegra _ _) =
+    error "TODO: makeSignedTransaction AllegraEra"
+makeSignedTransaction _ (ShelleyTxBody ShelleyBasedEraMary _ _) =
+    error "TODO: makeSignedTransaction MaryEra"
 
 makeByronKeyWitness :: NetworkId
                     -> TxBody ByronEra
                     -> SigningKey ByronKey
                     -> Witness ByronEra
+makeByronKeyWitness _ (ShelleyTxBody era _ _) = case era of {}
```

Comment:
This is my first time encountering `EmptyCase`, but is there a benefit here as opposed to using `error`?

Reply:
Yes! The benefit is that the compiler can tell you that it is correct!

With `error` you're promising the compiler that this can never happen. With empty case the compiler check for us that it can never happen. We have pattern match warnings turned on (and as errors in CI) and so the compiler can check that we're right that there are no cases here that we have missed. There are simply no possible cases at all.